app/instrumentation-tests/src/org/commcare/utils/CommCareIdlingResource.kt

```diff
@@ -0,0 +1,55 @@
+package org.commcare.utils
+
+import android.os.Handler
+import android.os.Looper
+import androidx.test.espresso.IdlingResource
+import androidx.test.platform.app.InstrumentationRegistry
+import org.commcare.CommCareInstrumentationTestApplication
+import org.commcare.activities.CommCareActivity
+
+/**
+ * An implementation of idling resource useful for monitoring idleness of progress dialogs.
+ *
+ * It has an additional idleness constraint that the progress dialog must not be shown for a set
+ * period of time before we declare the resource idle.
+ *
+ * Useful for cases where we make multiple network calls in succession where each response triggers
+ * another request until loading is complete.
+ * So the app will be like Not-Idle(Progress bar showing) -> Idle(Progress bar closed) -> Not-Idle -> Idle...
+ * And we don't want to report idle when this happens.
+ */
+class CommCareIdlingResource: IdlingResource {
+
+    companion object {
+        const val timeoutMs = 1000L
+    }
+
+    lateinit var resourceCallback: IdlingResource.ResourceCallback
+    var isIdle = false
+    val handler = Handler(Looper.getMainLooper())
+
+    override fun getName(): String {
+        return CommCareIdlingResource::class.java.name
+    }
+
+    override fun isIdleNow(): Boolean {
+        val activity = getCommCareActivity() ?: return true
+        if (!(activity.currentProgressDialog != null && activity.currentProgressDialog.isAdded)) {
```

Comment:
so effectlvely what we are trying to do is we are marking the resource as `Idle` a sec after the progress dialog has been removed. What's the behaviour if we don't add this at all ? The app never goes to `Idle` or takes more time to do so ?

Reply:
The app would take 1 more sec before going to idle.

Reply:
got it. thanks!

Reply:
might be worth adding that as comment either in the class javadoc or inline comment as that's the crux of why we need this class.

Reply:
Ooops Sorry I missed this comment before merging. 
BTW, I've already added docs here https://github.com/dimagi/commcare-android/pull/2305/files#diff-b30c4312c3ff1a1c9e97d4a741c20500R10-R20